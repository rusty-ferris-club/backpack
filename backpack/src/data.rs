use anyhow::Result as AnyResult;
use core::fmt::Debug;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Clone, Debug)]
pub struct Location {
    pub url: String,
    pub domain: String,
    pub path: String,
    pub project: String,
    pub port: Option<u16>,
    pub query: Option<String>,
    pub subfolder: Option<String>,
    pub gref: Option<String>,
    pub username: String,
    pub password: Option<String>,
    pub is_git: bool,
}
impl Location {
    pub fn git_url(&self) -> String {
        format!(
            "git@{}:{}.git",
            self.domain,
            self.path.trim_start_matches('/')
        ) // needs: auth, port
    }

    pub fn web_url(&self) -> String {
        format!("https://{}{}", self.domain, self.path) // needs: auth, port
    }

    /// Create a logical `Location` from a `Url`.
    ///
    /// # Errors
    ///
    /// This function will return an error if `url` is illegal
    pub fn from(url: &Url, is_git: bool) -> AnyResult<Self> {
        let path = url.path();
        let parts = path.split("/-/").collect::<Vec<_>>();
        let (path, subfolder) = if parts.len() == 2 {
            (parts[0], Some(parts[1].to_string()))
        } else {
            (path, None)
        };

        let project = path
            .split('/')
            .last()
            .ok_or_else(|| anyhow::anyhow!("cannot find project name"))?;

        Ok(Self {
            url: url.to_string(),
            domain: url.domain().unwrap_or_default().to_string(),
            path: path.trim_end_matches(".git").to_string(),
            project: project.to_string(),
            port: url.port(),
            query: url.query().map(str::to_string),
            subfolder,
            gref: url.fragment().map(str::to_string),
            username: url.username().to_string(),
            password: url.password().map(str::to_string),
            is_git,
        })
    }
}

#[derive(Debug)]
pub struct Archive {
    pub url: String,
    pub root: ArchiveRoot,
}

#[derive(Debug)]
pub enum ArchiveRoot {
    Folder(String),
    FirstFolder,
    None,
}
#[derive(Debug)]
pub struct Assets {
    pub archive: Option<Archive>,
    pub git: Option<String>,
}

pub enum Overwrite {
    Ask,
    Always,
    Never,
    Custom(Box<dyn Fn(&str) -> bool>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum CopyMode {
    #[serde(rename = "new")]
    Copy,
    #[serde(rename = "apply")]
    Apply,
    #[serde(rename = "all")]
    All,
}

#[allow(clippy::struct_excessive_bools)]
#[derive(Debug)]
pub struct Opts {
    pub show_progress: bool,
    pub overwrite: bool,
    pub mode: CopyMode,
    pub is_git: bool,
    pub no_cache: bool,
    pub always_yes: bool,
    pub remote: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_location_urls() {
        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            true,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc =
            Location::from(&Url::parse("https://github.com/user/repo").unwrap(), false).unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            false,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");
    }
}
