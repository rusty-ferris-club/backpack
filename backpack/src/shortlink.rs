use crate::{
    config::{Config, Swap},
    data::{Assets, Location},
    git::GitProvider,
    vendors::{Vendor, Vendors},
};
use anyhow::Result as AnyResult;
use interactive_actions::data::Action;
use lazy_static::lazy_static;
use regex::Regex;
use tracing;
use url::Url;

lazy_static! {
    static ref RE_GIT: Regex = Regex::new(r"^git@(.+?):(.+)$").unwrap();
    static ref RE_VENDOR: Regex = Regex::new(r"^([a-zA-Z0-9_-]+):(.+)$").unwrap();
}

fn expand<'a>(
    shortlink: &str,
    is_git: bool,
    vendors: &'a Vendors<'_>,
) -> AnyResult<(Box<dyn Vendor>, Location)> {
    let (vendor, url) = if shortlink.starts_with("https://") {
        //https://github.com/jondot/hygen/-/foobar

        let url = Url::parse(shortlink)?;
        let vendor = vendors.lookup(
            url.domain()
                .ok_or_else(|| anyhow::anyhow!("domain is missing"))?,
        )?;
        (vendor, url)
    } else if let Some(caps) = RE_GIT.captures(shortlink) {
        let domain = caps
            .get(1)
            .ok_or_else(|| anyhow::anyhow!("parse failed: no domain"))?
            .as_str();
        (
            vendors.lookup(domain)?,
            Url::parse(&format!(
                "https://{}/{}",
                domain,
                caps.get(2)
                    .ok_or_else(|| anyhow::anyhow!("parse failed: no path"))?
                    .as_str()
            ))?,
        )
    } else if let Some(caps) = RE_VENDOR.captures(shortlink) {
        let vendor = vendors.lookup(
            caps.get(1)
                .ok_or_else(|| anyhow::anyhow!("parse failed: no vendor"))?
                .as_str(),
        )?;
        let url = Url::parse(&format!(
            "https://{}/{}",
            vendor.base(),
            caps.get(2)
                .ok_or_else(|| anyhow::anyhow!("parse failed: no path"))?
                .as_str()
        ))?;

        (vendor, url)
    } else {
        let vendor = vendors.lookup("")?;
        let url = Url::parse(&format!("https://{}/{}", vendor.base(), shortlink))?;
        (vendor, url)
    };

    let location = Location::from(&url, is_git)?;
    Ok((vendor, location))
}

pub struct Shortlink<'a> {
    config: &'a Config,
    git: &'a dyn GitProvider,
}

impl<'a> Shortlink<'a> {
    pub fn new(config: &'a Config, git: &'a dyn GitProvider) -> Self {
        Self { config, git }
    }

    #[tracing::instrument(name = "shortlink_resolve", skip_all, err)]
    pub fn resolve(&self, shortlink: &str, is_git: bool) -> AnyResult<(Location, Assets)> {
        // try to get better settings from projects config:
        let (shortlink, is_git) = self.config.project(shortlink).map_or_else(
            || (shortlink, is_git),
            |project| (project.shortlink.as_str(), project.is_git.unwrap_or(false)),
        );

        // expand and resolve
        let vendors = Vendors::new(self.config.vendors.as_ref());
        let (vendor, location) = expand(shortlink, is_git, &vendors)?;
        let (location, assets) = vendor.resolve(&location, self.git)?;
        Ok((location, assets))
    }

    pub fn actions(&self, shortlink: &str) -> Option<&'a Vec<Action>> {
        self.config
            .project(shortlink)
            .and_then(|project| project.actions.as_ref())
    }
    pub fn swaps(&self, shortlink: &str) -> Option<&'a Vec<Swap>> {
        self.config
            .project(shortlink)
            .and_then(|project| project.swaps.as_ref())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    use insta::assert_debug_snapshot;

    macro_rules! set_snapshot_suffix {
        ($($expr:expr),*) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_prepend_module_to_snapshot(false);
            settings.set_snapshot_suffix(format!($($expr,)*));
            let _guard = settings.bind_to_scope();
        }
    }

    #[rstest]
    fn test_custom_config(
        #[values("", "gl:", "gh:", "ghe:", "notfound:")] vendor: &str,
        #[values("hygen", "jondot/hygen", "rust-starter")] slug: &str,
        #[values("")] gref: &str,
    ) {
        set_snapshot_suffix!(
            "[{}]-[{}]-[{}]",
            vendor.replace(':', ""),
            slug.replace('/', "_"),
            gref.replace('/', "_")
        );
        let config = Config::from_text(
            r###"
projects:
  rust-starter:
    shortlink: jondot/rust-starter
    is_git: true

vendors:
  default: 
    kind: gitlab
    base: my.gitlab.com

  custom:
    gh:
      kind: github
      base: beta.github.com/my-org
    ghe:
      kind: github
      base: github.enterprise.example.com
"###,
        )
        .unwrap();

        let link = format!("{}{}{}", vendor, slug, gref);
        let vendors = Vendors::new(config.vendors.as_ref());
        assert_debug_snapshot!((link.clone(), expand(&link, false, &vendors)));
    }

    #[rstest]
    fn test_smoke(
        #[values("", "gl:")] vendor: &str,
        #[values("jondot/hygen/-/subfolder")] slug: &str,
        #[values("", "#wip")] gref: &str,
    ) {
        set_snapshot_suffix!(
            "[{}]-[{}]-[{}]",
            vendor.replace(':', "_"),
            slug.replace('/', "_"),
            gref.replace('/', "_")
        );

        let link = format!("{}{}{}", vendor, slug, gref);
        let vendors = Vendors::new(None);
        assert_debug_snapshot!((link.clone(), expand(&link, false, &vendors)));
    }

    #[rstest]
    fn test_vendors(
        #[values(
            "",
            "gh:",
            "gl:",
            "bb:",
            "my-vendor:",
            "git@github.com:",
            "https://gitlab.com/"
        )]
        vendor: &str,
        #[values("jondot/hygen")] slug: &str,
        #[values("")] gref: &str,
    ) {
        set_snapshot_suffix!(
            "[{}]-[{}]-[{}]",
            vendor.replace(':', "_"),
            slug.replace('/', "_"),
            gref.replace('/', "_")
        );

        let link = format!("{}{}{}", vendor, slug, gref);
        let vendors = Vendors::new(None);
        assert_debug_snapshot!((link.clone(), expand(&link, false, &vendors)));
    }

    #[rstest]
    fn test_locations(
        #[values("")] vendor: &str,
        #[values("hygen", "jondot/hygen", "group/team/repo", "jondot/hygen/-/subfolder")]
        slug: &str,
        #[values("", "#wip")] gref: &str,
    ) {
        set_snapshot_suffix!(
            "[{}]-[{}]-[{}]",
            vendor.replace(':', "_"),
            slug.replace('/', "_"),
            gref.replace('/', "_")
        );

        let link = format!("{}{}{}", vendor, slug, gref);
        let vendors = Vendors::new(None);
        assert_debug_snapshot!((link.clone(), expand(&link, false, &vendors)));
    }
}
