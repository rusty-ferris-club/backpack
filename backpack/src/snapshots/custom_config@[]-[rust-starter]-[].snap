---
source: backpack/src/shortlink.rs
expression: "(link.clone(), expand(&link, false, &vendors))"
---
(
    "rust-starter",
    Ok(
        (
            my.gitlab.com,
            Location {
                url: "https://my.gitlab.com/rust-starter",
                domain: "my.gitlab.com",
                path: "/rust-starter",
                project: "rust-starter",
                port: None,
                query: None,
                subfolder: None,
                gref: None,
                username: "",
                password: None,
                is_git: false,
            },
        ),
    ),
)
